src/norm/mongodb.nim

```diff
@@ -0,0 +1,727 @@
+# Array (heterogeneous lists not supported yet)
+# DBPointer
+# Binary
+# Timestamp
+#
+# NOT SUPPORTED (ignored):
+#
+# Regexp
+# JSCode
+# JSCodeWithScope
+# MaximumKey
+# MinimumKey
+#
+# From Nim Object to BSON
+#
+# | Type      | BSON          | Notes
+# | --------- | ------------- |---------------------
+# | float     | Double        | default float in Nim already 64-bit
+# | string    | String (UTF8) | string already UTF-8
+# | Oid       | Oid           | https://nim-lang.org/docs/oids.html; all-zeroes count as a MISSING entry; not a null
+# | bool      | Bool          | 
+# | timestamp | TimeUTC       | https://nim-lang.org/docs/times.html; 1970-01-01 00:00:00 counts as MISSING; not null
+# | Option[T] | Null          | Only fields of Option[T] can be null
+# | int       | Int32         |
+#
+# NOT YET SUPPORTED (compile-time error generated):
+#
+# | float32   | Double        | 
+# | float64   | Double        | default float in Nim already 64-bit
+# | (Object)  | Document      | it is not possible to a "Missing" state. The closest analogue is Option[ObjectType]
+# | seq[T]    | Array         | to support null, make it seq[Option[T]]
+# | int32     | Int32         |
+# | int64     | Int64         |
+# Tuples
+# Tables
+#
+
+import
+  strutils,
+  sequtils,
+  macros,
+  typetraits,
+  logging,
+  options,
+  tables,
+  oids,
+  times
+
+import
+  nimongo/bson,
+  nimongo/mongo
+
+import
+  rowutils,
+  objutils,
+  pragmas
+
+
+export
+  strutils,
+  sequtils,
+  macros,
+  logging,
+  options,
+  tables,
+  oids,
+  times
+
+export
+  bson,
+  mongo
+
+export
+  rowutils,
+  objutils,
+  pragmas
+
+# compile-time globals
+var
+  normObjectNamesRegistry {.compileTime.}: seq[string] = @[]
+
+const NORM_UNIVERSAL_TYPE_LIST* = @[
+  "float",
+  "string",
+  "Oid",
+  "bool",
+  "Time",
+  "int",
+  "Option[float]",
+  "Option[string]",
+  "Option[Oid]",
+  "Option[bool]",
+  "Option[Time]",
+  "Option[int]",
+  "seq[float]",
+  "seq[string]",
+  "seq[Oid]",
+  "seq[bool]",
+  "seq[Time]",
+  "seq[int]"
+]
+
+proc getCollectionName*(objRepr: ObjRepr): string =
+  ##[ Get the name of the DB table for the given object representation:
+  ``table`` pragma value if it exists or lowercased type name otherwise.
+  ]##
+
+  result = objRepr.signature.name.toLowerAscii()
+
+  for prag in objRepr.signature.pragmas:
+    if prag.name == "table" and prag.kind == pkKval:
+      return $prag.value
+
+
+proc getCollectionName*(T: typedesc): string =
+  ##[ Get the name of the DB table for the given type: ``table`` pragma value if it exists
+  or lowercased type name otherwise.
+  ]##
+
+  when T.hasCustomPragma(table): T.getCustomPragmaVal(table)
+  else: ($T).toLowerAscii()  
+
+
+var
+  allCollections*: seq[string] = @[]
+  normConnection*: string
+  normUser*: string
+  normPassword*: string
+  normDatabase*: string
+
+template genWithoutDb(newCollections: seq[string]): untyped {.dirty.} =

```

Comment:
What is the purpose of this “withoutDb” thing*

Reply:
In the original backends, `getWithDb` stands for “generate `withDb` macro code.” So this must be generating `withoutDb` macro? But it does something completely different, I can't quite see what exactly now.

Reply:
Ah, misunderstood the meaning. I thought it meant "with DB parameters". I will rename it. This is for objects that are not tables/collections but are part of a document (or map column if postgresql).